### PR TITLE
Add/item in list syntax

### PR DESCRIPTION
**Pull request by Chyroc**

### CODE REVIEW

**Comment by Chyroc** [comment]
after https://github.com/mattn/anko/pull/180

**Comment by mattn** [comment]
I wonder this should be expr independently. For example `1 + 1 in [1,2,3]` or `for a in [1,2,3]` will be expected as another expression.

**Comment by Chyroc** [comment]
do you mean
```
expr_slice :
	IDENT '[' expr ':' expr ']'
	{
		$$ = ast.SliceExpr{Value: &ast.IdentExpr{Lit: $1.Lit}, Begin: $3, End: $5}
	}
	| IDENT '[' expr ':' ']'
	{
		$$ = ast.SliceExpr{Value: &ast.IdentExpr{Lit: $1.Lit}, Begin: $3, End: nil}
	}
	| IDENT '[' ':' expr ']'
	{
		$$ = ast.SliceExpr{Value: &ast.IdentExpr{Lit: $1.Lit}, Begin: nil, End: $4}
	}
	| expr '[' expr ':' expr ']'
	{
		$$ = ast.SliceExpr{Value: $1, Begin: $3, End: $5}
	}
	| expr '[' expr ':' ']'
	{
		$$ = ast.SliceExpr{Value: $1, Begin: $3, End: nil}
	}
	| expr '[' ':' expr ']'
	{
		$$ = ast.SliceExpr{Value: $1, Begin: nil, End: $4}
	}

expr_item_in_list :
	expr IN IDENT
	{
		$$ = &ast.IncludeExpr{ItemExpr: $1, ListExpr: ast.SliceExpr{Value: &ast.IdentExpr{Lit: $3.Lit}, Begin: nil, End: nil}}
	}
	| expr IN expr
	{
		$$ = &ast.IncludeExpr{ItemExpr: $1, ListExpr: ast.SliceExpr{Value: $3, Begin: nil, End: nil}}
	}
	| expr IN expr_slice
	{
		$$ = &ast.IncludeExpr{ItemExpr: $1, ListExpr: $3}
	}
```

?

**Comment by MichaelS11** [comment]
What about a slice in a multidimensional slice? Or other complex in statements?

What about a test for something like for `a in 5 in [1, 2, 3} { return a }` ?

**Comment by Chyroc** [comment]
do anko need to support `a in map`?

**Comment by Chyroc** [comment]
@MichaelS11 do you mean `a in 5 in [1, 2, 3]`

**Comment by MichaelS11** [comment]
Yes.

I say do not support `a in map` since we have `a, ok = aMap["key"]`

**Comment by Chyroc** [comment]
You are right, we have indeed implemented a judgment on whether an element is in a map clock。

Then I implemented `item in list`, maybe you can take a look again, then make some comments?

@mattn @MichaelS11

File: vm/vmExpr.go [hunk 1]
```diff
@@ -732,6 +732,28 @@ func invokeExpr(expr ast.Expr, env *Env) (reflect.Value, error) {
...
+		if err != nil {
+			return nilValue, newError(e.ItemExpr, err)
+		}
+		listExpr, err := invokeExpr(&e.ListExpr, env)
+		if err != nil {
+			return nilValue, newError(&e.ListExpr, err)
+		}
+
+		if listExpr.Kind() != reflect.Slice && listExpr.Kind() != reflect.Array {
+			return nilValue, newStringError(e, "second argument must be slice or array; but get "+listExpr.Kind().String())
```
Comment (MichaelS11) [inline vm/vmExpr.go]: Maybe change "but get" to "but have"

File: vm/vmItemInList_test.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		// for i in list && item in list
+		{script: `list_of_list = [["a"]];for l in list_of_list { return "a" in l }`, runOutput: true},
+
+		// not slice or array
+		// todo: support `"a" in "aaa"` ?
+		{script: `"a" in "aaa"`, runError: fmt.Errorf("second argument must be slice or array; but get string")},
+		{script: `1 in 12345`, runError: fmt.Errorf("type int64 does not support slice operation")},
+
+		// a in item in list
+		{script: `"a" in 5 in [1, 2, 3]`, runError: fmt.Errorf("type bool does not support slice operation")},
```
[the remark this answers is not in the document]
Comment (MichaelS11) [reply vm/vmItemInList_test.go]: Also `1++ in [1, 2, 3]` and `1 in 1++`

[the remark this answers is not in the document]
Comment (Chyroc) [reply vm/vmItemInList_test.go]: done

**Comment by Chyroc** [comment]
@MichaelS11 

should `1++ in [1, 2, 3]` and `1 in 1++` work? or return `Run error - received: Invalid operation`
```
{script: `1++ in [1, 2, 3]`, runOutput: true},
{script: `3++ in [1, 2, 3]`, runOutput: false},
{script: `1 in 1++`, runOutput: false},
```

vs


```
{script: `1++ in [1, 2, 3]`, runError: fmt.Errorf("Invalid operation")},
{script: `3++ in [1, 2, 3]`,runError: fmt.Errorf("Invalid operation")},
{script: `1 in 1++`, runError: fmt.Errorf("Invalid operation")},
```

**Comment by MichaelS11** [comment]
Testing is much better. Thank you. 

What about a slice in a multidimensional slice? Or other complex in statements?

[1] in [[1], [2], [3]]
[[1]] in [1,2,3]
etc...

**Comment by Chyroc** [comment]
@MichaelS11 add test
```
		{script: `1 in [1]`, runOutput: true},
		{script: `1 in [[1]]`, runOutput: false},
		{script: `1 in [[[1]]]`, runOutput: false},
		{script: `1 in [[1],[[1]],1]`, runOutput: true},
		{script: `[1] in [1]`, runOutput: false},
		{script: `[1] in [[1]]`, runOutput: true},
		{script: `[1] in [[[1]]]`, runOutput: false},
		{script: `[[1]] in [1]`, runOutput: false},
		{script: `[[1]] in [[1]]`, runOutput: false},
		{script: `[[1]] in [[[1]]]`, runOutput: true},
```

**Comment by MichaelS11** [comment]
Great. If wanted, could also add all the tests so far for data on the outside (using input), plus a mixing of the two (VM data plus input data).

**Comment by Chyroc** [comment]
I'm sorry, my English is not very good, so I'll confirm it again.
What you mean is to save the test(`TestItemInList`) as a file in the `t` directory (test the anko code read from the file by the anko binaty) and do the test?
As well as mixing reads from files by anko binary, and calling `vm.Run` directly?

**Comment by MichaelS11** [comment]
For example:

Here is tests with VM maps
https://github.com/mattn/anko/blob/master/vm/vmArraysAndMaps_test.go#L732-L737

Here is tests with outside/Go maps:
https://github.com/mattn/anko/blob/master/vm/vmArraysAndMaps_test.go#L769-L776

**Comment by Chyroc** [comment]
@MichaelS11 equal function in `vm/vm.go` convert string "1" to int 1, so "1" == 1，it is expected?

and should it right:
```
"1" == 1
1 == "1"
"1" in [1]
1 in ["1"]
```
?

**Comment by Chyroc** [comment]
@MichaelS11 add some test with input, need review~

File: vm/vmExpr.go [hunk 3]
```diff
@@ -756,7 +756,23 @@ func invokeExpr(expr ast.Expr, env *Env) (reflect.Value, error) {
...
+			list := listExpr.Index(i)
+
+			if item.Kind() == reflect.Interface || item.Kind() == reflect.Ptr {
+				item = item.Elem()
+			}
+			if list.Kind() == reflect.Interface || list.Kind() == reflect.Ptr {
+				list = list.Elem()
+			}
+
+			if isNum(item) && isNum(list) {
```
Comment (mattn) [inline vm/vmExpr.go]: Just questino. Why only number?

Comment (Chyroc) [reply vm/vmExpr.go]: if not this, than 
int(1) != int64(1) when use input, etc...

Comment (Chyroc) [reply vm/vmExpr.go]: ok~

**Comment by MichaelS11** [comment]
Could you help me understand why equalNumber is needed? Is there a reason why the equal function is not being used?

**Comment by Chyroc** [comment]
@MichaelS11 

use equal, than 
```
1 == 1.1
1.1 != 1
```
This doesn't work, this is a bug, if it's just that, I can use equal and than fix the equal

but :
```
"1" == 1
"true" == true
```
This seems to be the functional point of equal, but this is not intuition

**Comment by MichaelS11** [comment]
Please use equal then and we can working on addressing any bugs.

I say match JavaScript equal: 
```
console.log(1 == 1.1) // false
console.log(1 != 1.1) // true
console.log("1" == 1) // true
console.log("true" == true) // false
```

**Comment by Chyroc** [comment]
@MichaelS11 i replace equalNumber by equal,

**Comment by MichaelS11** [comment]
LGTM

**Comment by mattn** [comment]
Thank you

**Pull request merged**